### PR TITLE
add Samsung Galaxy Tab trackpad gestures

### PR DESCRIPTION
**Pull request by AlphaBs**
add some gestures of Samsung Galaxy Tab trackpad (SEC-EFDT970).
- tap to left click
- long tap to drag
- two-finger tap to right click
- two-finger scroll

tested on Samsung Galaxy Tab S8+ (SM-X800), Android 14, OneUI 6.0

I also recorded `MotionEvent`s when I gesture:
```
move
action: ACTION_HOVER_MOVE, flags: 40, state: 0

tap
action: ACTION_HOVER_EXIT, flags: 40, state: 0 
action: ACTION_DOWN, flags: 0, state: 0
action: ACTION_UP, flags: 40, state: 0
action: ACTION_HOVER_ENTER, flags: 40, state: 0

long tap
action: ACTION_HOVER_EXIT, flags: 40, state: 0
action: ACTION_DOWN, flags: 0, state: 0
action: ACTION_MOVE, flags: 0, state: 0 <- drag
action: ACTION_MOVE, flags: 2000000, state: 0 <- randomly
action: ACTION_UP, flags: 40, state: 0
action: ACTION_HOVER_ENTER, flags: 40, state: 0

two-finger tap
action: ACTION_HOVER_EXIT, flags: 40, state: 0
action: ACTION_DOWN, flags: 40, state: 2
action: ACTION_BUTTON_PRESS, flags: 0, state: 2
action: ACTION_BUTTON_RELEASE, flags: 0, state: 0
action: ACTION_UP, flags: 40, state: 0
action: ACTION_HOVER_ENTER, flags: 40, state: 0

two-finger scroll
action: ACTION_HOVER_EXIT, flags: 40, state: 0
action: ACTION_DOWN, flags: 10000040, state: 0
action: ACTION_MOVE, flags: 10000040, state: 0 <- scroll
action: ACTION_MOVE, flags: 12000040, state: 0 <- randomly
action: ACTION_UP, flags: 40, state: 0
action: ACTION_HOVER_ENTER, flags: 40, state: 0

press left button
action: ACTION_HOVER_EXIT, flags: 40, state: 0
action: ACTION_DOWN, flags: 0, state: 1
action ACTION_BUTTON_PRESS, flags: 0, state: 1
action ACTION_BUTTON_RELEASE, flags: 0, state: 0
action ACTION_UP, flags: 40, state: 0
action ACTION_HOVER_ENTER, flags: 40, state: 0
action ACTION_HOVER_EXIT, flags: 40, state: 0 <- hand off

press right button
action: ACTION_HOVER_EXIT, flags: 40, state: 0
action: ACTION_DOWN, flags: 0, state: 2
action: ACTION_BUTTON_PRESS, flags: 0, state: 2
action: ACTION_BUTTON_RELEASE, flags: 0, state: 0
action: ACTION_UP, flags: 40, state: 0
action ACTION_HOVER_ENTER, flags: 40, state: 0
action: ACTION_HOVER_EXIT, flags: 40, state: 0
```

### CODE REVIEW

**Comment by twaik** [comment]
I did not dive deep to changes but I already can say that kt will break support for other dex devices. Replacing one constant (`0x14000000`) with another one (`0x10000000`) is not an option.
Not going to be merged in the current state.

**Comment by AlphaBs** [comment]
okay do we need completely different way to detect gestures? or just `hasFlags(e, 0x10000000) || hasFlags(e, 0x14000000)` is enough?

**Comment by AlphaBs** [comment]
I reverted the constant

**Comment by twaik** [comment]
Ok, I need to test this on my device and request test from one other guy.

File: app/src/main/java/com/termux/x11/input/TouchInputHandler.java [hunk 1]
```diff
@@ -595,6 +595,10 @@ private boolean hasFlags(MotionEvent e, int flags) {
             return (e.getFlags() & flags) == flags;
         }
 
+        private boolean isScrollingEvent(MotionEvent e) {
+            return hasFlags(e, 0x14000000) || e.getClassification() == MotionEvent.CLASSIFICATION_TWO_FINGER_SWIPE;
```
Comment (twaik) [inline app/src/main/java/com/termux/x11/input/TouchInputHandler.java:599]:
```suggestion
return hasFlags(e, 0x14000000) || (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && e.getClassification() == MotionEvent.CLASSIFICATION_TWO_FINGER_SWIPE);
```

Call requires API level 29 (current min is 26).

**Comment by twaik** [comment]
Ok, it works fine on Samsung Galaxy S20 FE (SM-G780F).

**Comment by twaik** [comment]
Can you please investigate why it does not work with `Capture external mouse when possible`?

**Comment by AlphaBs** [comment]
Sure, but my device behaves very strangely (inversed X/Y axis, can't use any gestures) when I turn that option on, even before the PR code.

**Comment by twaik** [comment]
AFAIK it should be handled by `HardwareMouseListener`. And I am not sure why it inverts X/Y axis, I probably will implement options like `External mouse rotation` to сompensate this. But gestures are still unhandled and I do not really have resources to fix this.

File: app/src/main/java/com/termux/x11/input/TouchInputHandler.java [hunk 2]
```diff
@@ -21,6 +22,8 @@
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
 
+import android.util.Log;
```
Comment (twaik) [inline app/src/main/java/com/termux/x11/input/TouchInputHandler.java:25]:
```suggestion
```
Probably this line should be removed.

**Comment by AlphaBs** [comment]
I'm just curious, why trackpad need captured pointer? I'm not sure of the intention behind this feature.

**Comment by twaik** [comment]
The "Captured Pointer" feature offers several advantages. Firstly, the navigation and status bars do not appear when the mouse touches the top or bottom edges of the screen. Secondly, Android sends relative mouse events instead of the absolute ones typical in regular mode. This is particularly important in scenarios like 3D games or 3D editors, or in any other situation where an X11 application manipulates the cursor's position, and this behavior needs to be taken into account.

**Comment by AlphaBs** [comment]
in my device, just removing below code fixes the issue. 

TouchInputHandler.java:
```java 
// Regular touchpads and Dex touchpad send events as finger too,
// but they should be handled as touchscreens with trackpad mode.
if  (mTouchpadHandler != null && (event.getSource() & InputDevice.SOURCE_TOUCHPAD) == InputDevice.SOURCE_TOUCHPAD) {
    return mTouchpadHandler.handleTouchEvent(view, view, event);
}
```

With trackpad mode, all gestures including tap, long tap to drag, two-finger tap, two-finger scroll, even three-finger tap are handled well. 

but X/Y axis are still ininverted. It looks like Android or my trackpad is reporting the wrong MotionEvent. As you said we might need an option to invert this.

**Comment by AlphaBs** [comment]
`MotionEvent` from my touchpad with captured pointer should be handled by `mInputStrategy`. I don't know why it was handled by `mHMListener`. Maybe there was an another touchpad that should be handled by `HardwareMouseListener`? I don't have another touchpad I can't test futher.

**Comment by twaik** [comment]
There is no `two-finger drag` feature in the original code. I will not merge it.

**Comment by twaik** [comment]
Or it is only for the case of Dex? Yeah, it seems like yes.

**Comment by AlphaBs** [comment]
My device supports two-finger dragging. I can use this gesture in any android app, not just termux-x11. 
This gesture is useful when I need to drag a very long distance accurately. 
And I was able to achieve this by simply keeping the `onTap` state when ACTION_BUTTON_PRESS. (the real event is `ACTION_HOVER_EXIT -> ACTION_DOWN -> ACTION_BUTTON_PRESS -> ACTION_MOVE(dragging) -> ACTION_BUTTON_RELEASE -> ACTION_UP`)
Is this a potential backwards compatibility breaker?

**Comment by twaik** [comment]
I am not really sure about this. Need some check.

**Comment by AlphaBs** [comment]
> Or it is only for the case of Dex? Yeah, it seems like yes.

yes! Is there anything else I need to do before it merges?

**Comment by twaik** [comment]
Currenly I have to do some checks on real device.

**Comment by twaik** [comment]
Probably that is safe, but I can not reproduce two-finger drag behaviour.

**Pull request merged**